tests/params/test_tensor.py

```diff
@@ -0,0 +1,243 @@
+    the methods return the expected result for different second order permeability
+    tensors. Specifically we consider here isotropic, anisotropic, diagonal,
+    non-diagonal, homogeneous and heterogeneous tensors. All three spatial dimensions
+    are tested.
+
+    Parameters:
+        kxx: Array with cell-wise values of kxx permeability.
+        kyy: Array of kyy.
+        kzz: Array of kzz.
+        kxy: Array of kxy.
+        kyz: Array of kyz.
+
+    """
+    kwargs = {"kxx": kxx}
+    if kyy is not None:
+        kwargs["kyy"] = kyy
+    if kxy is not None:
+        kwargs["kxy"] = kxy
+    if kyz is not None:
+        kwargs["kyz"] = kyz
+
+    perm = pp.SecondOrderTensor(**kwargs)
+
+    assert perm.is_diagonal(nd) == expected_diagonal
+    assert perm.is_isotropic(nd) == expected_isotropic
+
+
+@pytest.mark.parametrize(
+    "second_order",
+    [(True), (False)],
+)
+def test_restrict_to_cells(second_order):
+    """Testing the restriction of a tensor to a chosen set of cells.
+
+    The method restrict_to_cells is used to restrict e.g. the permeability tensor to a
+    region within the simulation domain. This method checks that this restriction is
+    done correctly.
+
+    """
+    num_cells = 16
+    if second_order:
+        np.random.seed(42)
+
+        # Defining an arbitrary anisotropic second order tensor:
+        kxx = np.random.uniform(0.5, 1.0, num_cells)
+        kyy = np.random.uniform(0.5, 1.0, num_cells)
+        kxy = np.random.uniform(0.1, 0.5, num_cells)
+
+        tensor = pp.SecondOrderTensor(kxx=kxx, kyy=kyy, kxy=kxy)
+    else:
+        tensor = fourth_order_tensor_for_testing(
+            num_cells=num_cells, return_tensor_and_matrix=False
+        )
+    # Define some cells which we want to restrict the tensor k to:
+    active_cells = np.array([2, 1, 14, 10, 15])
+
+    # First copy k, and then restrict the copied k to the cells we chose above.
+    tensor_copy = tensor.copy()
+    tensor_copy.restrict_to_cells(active_cells)
+
+    # Looping through the permeability tensors to check that the restriction is done
+    # correctly.
+    for ind, cell in enumerate(active_cells):
+        original = tensor.values[:, :, cell]
+        copied_restriction = tensor_copy.values[:, :, ind]
+
+        assert np.array_equal(original, copied_restriction)
+
+
+def test_custom_field_tensor_generation():
+    """Test generation of tensors based on custom fields.
+
+    Utilizes the function fourth_order_tensor_for_testing() to generate a tensor with
+    custom fields. The tensor is assumed to be homogeneous, and the function returns
+    both the tensor and a hardcoded matrix representation of the tensor for one cell.
+    The matrix representation is used as a blueprint for checking if the tensor is
+    constructed correctly.
+
+    """
+    tensor, matrix = fourth_order_tensor_for_testing(
+        num_cells=5, return_tensor_and_matrix=True
+    )
+    num_cells = tensor.values.shape[2]
```

Comment:
Rather make `num_cells` a parameter, feed it to the factory method, and check that  `shape[2]` is as expected. It will always be, but why not test it.

Reply:
It is not clear to me what you mean here. Can you please elaborate?

Reply:
Never mind, I got it, I just misread your comment. Thanks!